fasm2bels/lib/logical_netlist.py

```diff
@@ -0,0 +1,274 @@
+        assert port in self.ports
+        port = PortInstance(name=port, idx=idx, instance_name=None)
+        self.nets[net_name].ports.append(port)
+
+
+def invert_direction(direction):
+    """ Inverts direction of Direction enum. """
+    if direction == Direction.Input:
+        return Direction.Output
+    elif direction == Direction.Output:
+        return Direction.Input
+    else:
+        assert direction == Direction.Inout
+        return Direction.Inout
+
+
+class Library():
+    """ Library of cells. """
+
+    def __init__(self, name):
+        self.name = name
+        self.cells = {}
+
+    def add_cell(self, cell):
+        assert cell.name not in self.cells
+        self.cells[cell.name] = cell
+
+
+def check_logical_netlist(libraries):
+    """ Check that a logical netlist is consistent.
+
+    The following things are checked:
+     - All cell instances have a paired cell.
+     - All cells have a unique cell definition.
+     - All port connections on nets match the paired cell model.
+     - Nets without Inout ports have only 1 driver.
+
+    Returns a set of Cell names.
+
+    """
+    master_cell_list = {}
+
+    for lib in libraries.values():
+        for cell in lib.cells.values():
+            assert cell.name not in master_cell_list
+            master_cell_list[cell.name] = cell
+
+    for cell in master_cell_list.values():
+        for inst in cell.cell_instances.values():
+            assert inst.cell_name in master_cell_list
+
+        for netname, net in cell.nets.items():
+            port_directions = {
+                Direction.Input: 0,
+                Direction.Output: 0,
+                Direction.Inout: 0,
+            }
+
+            for port in net.ports:
+                if port.instance_name is not None:
+                    # This port connects to a cell instance, go find the
+                    # master cell and port.
+                    instance_cell_name = cell.cell_instances[
+                        port.instance_name].cell_name
+                    instance_cell = master_cell_list[instance_cell_name]
+                    instance_port = instance_cell.ports[port.name]
+
+                    net_direction = invert_direction(instance_port.direction)
+                else:
+                    instance_port = cell.ports[port.name]
+
+                    net_direction = instance_port.direction
+
+                # Count port directions on this net
+                port_directions[net_direction] += 1
+
+                # Check bus index is valid is present
+                if port.idx is not None:
+                    assert instance_port.bus is not None
+
+                    assert port.idx >= instance_port.bus.start
+                    assert port.idx <= instance_port.bus.end
+                else:
+                    assert instance_port.bus is None, (netname, port)
+
+            if port_directions[Direction.Inout] == 0:
+                assert port_directions[Direction.Input] in [
```

Comment:
Shouldn't there be a check if output count > 0? A net must have a driver.

Reply:
Partial logical netlists are allowed in the interchange, and so I didn't assert that it had to have a driver, only that it cannot have multiple drivers.

Reply:
Ok